libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/OAuth2.java

```diff
@@ -152,13 +157,34 @@ public static URI getAuthorizationUrl(URI loginServer, String clientId,
     
     public static URI getAuthorizationUrl(URI loginServer, String clientId,
             String callbackUrl, String[] scopes, String clientSecret, String displayType) {
+        return URI.create(getAuthorizationUrlStr(loginServer, clientId, callbackUrl, scopes, clientSecret, displayType));
+    }
+
+
+    public static URI getAuthorizationUrl(URI loginServer, String clientId,
+                                          String callbackUrl, String[] scopes, String clientSecret,
+                                          String displayType, String accessToken, String instanceURL) {
+        if(accessToken == null || instanceURL == null) {
+            return getAuthorizationUrl(loginServer, clientId, callbackUrl, scopes, clientSecret, displayType);
+        }
+        final StringBuilder sb = new StringBuilder(instanceURL);
+        sb.append(FRONTDOOR);
+        sb.append(SID).append(EQUAL).append(accessToken);
+        sb.append(AND).append(RETURL).append(EQUAL).append(Uri.encode(getAuthorizationUrlStr(loginServer,clientId,callbackUrl, scopes, clientSecret, displayType)));
+        return URI.create(sb.toString());
+    }
+
+
+    private static String getAuthorizationUrlStr(URI loginServer, String clientId,
```

Comment:
Why do we need two methods? Couldn't we just have the one returning a URI, and if a string is wanted instead, the caller could call toString()?